trio/_core/_run.py

```diff
@@ -1118,6 +1130,62 @@ class Runner:
     entry_queue = attr.ib(factory=EntryQueue)
     trio_token = attr.ib(default=None)
 
+    # Guest mode stuff
+    is_guest = attr.ib(default=False)
+    run_sync_soon_threadsafe = attr.ib(default=None)
+    done_callback = attr.ib(default=None)
+    unrolled_run_gen = attr.ib(default=None)
+    unrolled_run_next_send = attr.ib(default=None)
+    guest_tick_scheduled = attr.ib(default=False)
+
+    def guest_tick(self):
+        # XX no signal handling support at all currently
+        assert self.is_guest
+        try:
+            timeout = self.unrolled_run_gen.send(self.unrolled_run_next_send)
+        except StopIteration:
+            GLOBAL_RUN_CONTEXT.__dict__.clear()
+            self.close()
+            # XX if we had KI support, we'd have to do something with it here
+            self.done_callback(self.main_task_outcome)
+            return
+
+        self.unrolled_run_next_send = None
+
+        # Optimization: do a zero-timeout check for already-pending I/O from
+        # the main thread
+        events = self.io_manager.get_events(0)
+        if events or timeout <= 0:
+            self.unrolled_run_next_send = events
+            self.guest_tick_scheduled = True
+            self.run_sync_soon_threadsafe(self.guest_tick)
+        else:
+            self.guest_tick_scheduled = False
+
+            def get_events():
+                return self.io_manager.get_events(timeout)
+
+            def deliver(events_outcome):
+                def in_main_thread():
+                    self.unrolled_run_next_send = events_outcome.unwrap()
+                    self.guest_tick_scheduled = True
+                    self.guest_tick()
+
+                self.run_sync_soon_threadsafe(in_main_thread)
+
+            # XX temporary placeholder until #1545 is merged
+            def start_thread_soon(d, fn):
+                t = threading.Thread(daemon=True, target=lambda: deliver(capture(fn)))
+                t.start()
+
+            start_thread_soon(deliver, get_events)
+
+    def force_guest_tick_asap(self):
```

Comment:
Reminder on this?

Reply:
I think I'm going to leave it as "asap". Normally we use `soon` to mean "hey remember that this might not happen until some arbitrary time in the future", and here we kind of mean the opposite: "it was going to happen later, but we can't wait! make it happen now!".